### PR TITLE
preprocessor, planner: inline processing by default for single-consumer cte

### LINKED ISSUE #36295: Planner should automatically inline CTE if it has only 1 consumer
## Enhancement

e.g.:
```sql
EXPLAIN WITH repo_stars AS (
    SELECT
       /*+ read_from_storage(tiflash[github_events]) */
        repo_id,
        ANY_VALUE(repos.name) AS repo_name,
        COUNT(distinct actor_login) AS stars
    FROM github_events
         JOIN db_repos repos ON repos.id = github_events.repo_id
    WHERE type = 'WatchEvent'
    GROUP BY 1
), top_10_repos AS (
    SELECT
        repo_id, repo_name, stars
    FROM repo_stars rs
    ORDER BY stars DESC
    LIMIT 10
)
SELECT repo_id, repo_name, stars from top_10_repos;

+----------------------------------------------------+---------------+--------------+--------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| id                                                 | estRows       | task         | access object                                          | operator info                                                                                                                                                                                                                                                                              |
+----------------------------------------------------+---------------+--------------+--------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| CTEFullScan_80                                     | 10.00         | root         | CTE:top_10_repos                                       | data:CTE_1                                                                                                                                                                                                                                                                                 |
| CTE_1                                              | 10.00         | root         |                                                        | Non-Recursive CTE                                                                                                                                                                                                                                                                          |
| └─TopN_68(Seed Part)                               | 10.00         | root         |                                                        | Column#43:desc, offset:0, count:10                                                                                                                                                                                                                                                         |
|   └─CTEFullScan_75                                 | 825.74        | root         | CTE:rs                                                 | data:CTE_0                                                                                                                                                                                                                                                                                 |
| CTE_0                                              | 825.74        | root         |                                                        | Non-Recursive CTE                                                                                                                                                                                                                                                                          |
| └─Projection_19(Seed Part)                         | 825.74        | root         |                                                        | gharchive_dev.github_events.repo_id, any_value(gharchive_dev.db_repos.name)->Column#40, Column#39                                                                                                                                                                                          |
|   └─TableReader_55                                 | 825.74        | root         | table of repos, partition:watch_event of github_events | data:ExchangeSender_54                                                                                                                                                                                                                                                                     |
|     └─ExchangeSender_54                            | 825.74        | mpp[tiflash] |                                                        | ExchangeType: PassThrough                                                                                                                                                                                                                                                                  |
|       └─Projection_53                              | 825.74        | mpp[tiflash] |                                                        | Column#39, gharchive_dev.github_events.repo_id, gharchive_dev.db_repos.name                                                                                                                                                                                                                |
|         └─HashAgg_21                               | 825.74        | mpp[tiflash] |                                                        | group by:gharchive_dev.github_events.repo_id, funcs:count(distinct gharchive_dev.github_events.actor_login)->Column#39, funcs:firstrow(gharchive_dev.github_events.repo_id)->gharchive_dev.github_events.repo_id, funcs:firstrow(gharchive_dev.db_repos.name)->gharchive_dev.db_repos.name |
|           └─ExchangeReceiver_52                    | 825.74        | mpp[tiflash] |                                                        |                                                                                                                                                                                                                                                                                            |
|             └─ExchangeSender_51                    | 825.74        | mpp[tiflash] |                                                        | ExchangeType: HashPartition, Hash Cols: [name: gharchive_dev.github_events.repo_id, collate: binary]                                                                                                                                                                                       |
|               └─Projection_49                      | 825.74        | mpp[tiflash] |                                                        | gharchive_dev.github_events.repo_id, gharchive_dev.github_events.actor_login, gharchive_dev.db_repos.name                                                                                                                                                                                  |
|                 └─HashJoin_50                      | 825.74        | mpp[tiflash] |                                                        | inner join, equal:[eq(Column#49, Column#48)]                                                                                                                                                                                                                                               |
|                   ├─ExchangeReceiver_33(Build)     | 46.00         | mpp[tiflash] |                                                        |                                                                                                                                                                                                                                                                                            |
|                   │ └─ExchangeSender_32            | 46.00         | mpp[tiflash] |                                                        | ExchangeType: Broadcast                                                                                                                                                                                                                                                                    |
|                   │   └─Projection_30              | 46.00         | mpp[tiflash] |                                                        | gharchive_dev.db_repos.name, cast(gharchive_dev.db_repos.id, double BINARY)->Column#49                                                                                                                                                                                                     |
|                   │     └─TableFullScan_31         | 46.00         | mpp[tiflash] | table:repos                                            | keep order:false, stats:pseudo                                                                                                                                                                                                                                                             |
|                   └─Projection_34(Probe)           | 315175383.76  | mpp[tiflash] |                                                        | gharchive_dev.github_events.repo_id, gharchive_dev.github_events.actor_login, cast(gharchive_dev.github_events.repo_id, double BINARY)->Column#48                                                                                                                                          |
|                     └─Selection_36                 | 315175383.76  | mpp[tiflash] |                                                        | eq(gharchive_dev.github_events.type, "WatchEvent")                                                                                                                                                                                                                                         |
|                       └─TableFullScan_35           | 4806644995.00 | mpp[tiflash] | table:github_events                                    | keep order:false, PartitionTableScan:true                                                                                                                                                                                                                                                  |
+----------------------------------------------------+---------------+--------------+--------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
21 rows in set (8.06 sec)
```

Every CTE has  only 1 consumer, we can inline them all.

### PR DESCRIPTION
**Pull request by elsa0520**
### What problem does this PR solve?

This pr mainly does an optimization of inline CTE.
When there is only one consumer of CTE, the CTE is inlined by default. (Except recursive CTE )

Issue Number: close #36295

Problem Summary:

### What is changed and how it works?

There are two changes:
1. Added the count processing of CTE consumers in AST perprocessor.
2. When the AST is converted into a logical plan, inline processing is performed for the CTE whose count is 1.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by winoros — COMMENTED** [review]
The pushed diff missed that push the new slice into the stack when we meet a new query block?

**Comment by elsa0520** [comment]
> The pushed diff missed that push the new slice into the stack when we meet a new query block?

Some code forgot to submit 😭

**Comment by elsa0520** [comment]
How to rerun check_dev_2  ？L
ooking at the error log, it seems to be a problem with the cluster itself, not pr.

**Comment by elsa0520** [comment]
> /run-check_dev_2

Thanks ~

**Review by fixdb — COMMENTED** [review]
Not sure about this, it should be safe to inline this, right?
```sql
with x as (select * from (select b from t for update) s)
select * from x where b = 1;
```

**Review by fixdb — APPROVED** [review]

**Comment by fixdb** [comment]
@elsa0520 issue https://github.com/pingcap/tidb/issues/36295

**Review by winoros — APPROVED** [review]

File: planner/core/preprocess.go [hunk 1]
```diff
@@ -570,6 +604,14 @@ func (p *preprocessor) Leave(in ast.Node) (out ast.Node, ok bool) {
 		if cteNode, exist := x.(*ast.CommonTableExpression); exist {
 			with.cteCanUsed = append(with.cteCanUsed, cteNode.Name.L)
 		}
+	case *ast.SelectStmt:
+		if x.With != nil {
+			p.preprocessWith.cteStack = p.preprocessWith.cteStack[0 : len(p.preprocessWith.cteStack)-1]
```
Comment (AilinKid) [inline planner/core/preprocess.go:609]: `[: len(p.preprocessWith.cteStack)-1]` can be this for simplicity

**Review by AilinKid — APPROVED** [review]

File: planner/core/preprocess.go (hunk 1 above)
Comment (elsa0520) [reply planner/core/preprocess.go:609]: Great ~ . I will fix it in the next time ~

**Pull request merged**